### ISSUE TITLE
`easystats` for ML

### ISSUE THREAD

**Issue by pdwaggoner** [easystats/performance#57]
Here is where we can more succinctly workshop an idea for approaching the ML world from an easystats perspective. Per @strengejacke 's suggestion, I wanted to just get the conversation started for a simple function idea to add to `performance` (based on `model_performance.lm`):

```{R}
# quick idea for cluster objects (kmeans, hierarchical, etc.)
model_performance.cluster <- function(d, x, metrics = "all", ...) {

  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if (!requireNamespace("factoextra", quietly = TRUE)) {
    stop("Package \"factoextra\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
    stop('"d" must be data frame\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
# requring an "eclust" object here, because the eclust function from factoextra allows for specifying many types of clustering algorithms (e.g., diana, pam, hierarchical, kmeans, etc.)
  x <- try(methods::as(x, "eclust", strict=TRUE))
  if (!inherits(x, "eclust")){
    stop('"x" must be a eclust object',
         'You have provided an object of class: ', class(x)[1])
  } 
  
  # calculate and store validation stats for cluster model
  cluster_stats <- fpc::cluster.stats(stats::dist(d), x$cluster)
  
  if (all(metrics == "all")) {
    metrics <- c("cluster.number", "cluster.size", "average.between", "average.within", 
                 "silhouette.widths", "within.cluster.ss", "dunn", "rand", "vi")
  }
  
  out <- list()
  
  # cluster.number: number of clusters
  if ("cluster.number" %in% metrics) {
    out$cluster.number <- cluster_stats$cluster.number
  }
  
  # cluster.size: number of obs in each cluster
  if ("cluster.size" %in% metrics) {
    out$cluster.size <- cluster_stats$cluster.size
  }
  
  # average.between: average distance between clusters
  if ("average.between" %in% metrics) {
    out$average.between <- cluster_stats$average.between
  }
  
  # average.within: average distance within clusters
  if ("average.within" %in% metrics) {
    out$average.within <- cluster_stats$average.within
  }
  
  # clus.avg.silwidths: average silhouette widths
  if ("silhouette.widths" %in% metrics) {
    out$silhouette.widths <- cluster_stats$clus.avg.silwidths
  }
  
  # within.cluster.ss: within clusters sum of squares
  if ("within.cluster.ss" %in% metrics) {
    out$within.cluster.ss <- cluster_stats$within.cluster.ss
  }
  
  # dunn: Dunn index
  if ("dunn" %in% metrics) {
    out$dunn <- cluster_stats$dunn
  }
 
  # corrected.rand: (corrected) Rand index assesses cluster similarity
  if ("rand" %in% metrics) {
    out$rand <- cluster_stats$corrected.rand
  }
  
  #vi: Meila’s variation of information (VI) index assesses cluster similarity
  if ("vi" %in% metrics) {
    out$vi <- cluster_stats$vi
  }
  
  out <- as.data.frame(out)
  row.names(out) <- NULL
  out
}
```

Is this more in line with something you're considering? If not, let me know as there are many directions we could take this. 

Re: `caret`, I think it would take a bit more careful thought, as with most of the more advanced ML techniques, as there is much more behind "fitting" a model, such as tuning algorithms, creating and iterating over training and testing data sets, and so on. Its significantly more involved (though not overly arduous) than fitting a basic `lm` or `glm`, for example. Let me know your thoughts here, and we can decide how to best move forward in the ML world as it makes sense with the mission and focus of the `easyverse` (very nice, by the way ;) , love it)

**Comment by strengejacke** [comment]
@DominiqueMakowski Just revisiting this issue again... We have some cluster-stuff in _parameters_. I'm not sure if it makes sense to put parts of the code into that package, and leave some parts related to model performance here?

**Comment by pdwaggoner** [comment]
Yeah, having code in separate places is a bit odd. However, given that there is no current substantive support for ML methods (whether unsupervised like clustering or otherwise), perhaps it makes most sense as is now, and then possibly as we develop support for ML, we can revisit this and consolidate into a new easystats package, `Machine Learning`, or something like that. Not sure if that makes any sense from your perspective/vision?

**Comment by DominiqueMakowski** [comment]
I am not sure having a dedicated package for ML would conceptually be meaningful, as in the end, ML is just a fancy name for predictive modelling 😁, parts of which are taken care of by performance, parameters and all (I mean "performance" and "parameters" have a scope large enough to describe parameters and performance from regression and why not neural nets and all)

However, some aspects often used in ML, such as feature reduction/selection, could indeed require a special treatment -> https://github.com/easystats/easystats/issues/47

**Comment by pdwaggoner** [comment]
@DominiqueMakowski fair enough re: having certain packages already taking care of some things related to ML. However, I'd disagree that ML is a fancy name for predictive modeling. Though, yes, forecasting and prediction are key goals of ML, so too is classification. Further, ML is premised on a fundamentally different way of building models, from train/test splits to numeric optimization techniques, all for the goal of building effective learners based on historical patterns. See the canonical paper on this treatment, for example, https://kcir.pwr.edu.pl/~witold/ai/cacm12.pdf, for whatever it might be worth.

Yet, ultimately, from an `easystats` perspective, I agree that perhaps it doesn't make conceptual sense to build out the ML side of things so explicitly. Thanks for the dialogue!

**Comment by DominiqueMakowski** [comment]
> Further, ML is premised on a fundamentally different way of building models, from train/test splits to numeric optimization techniques

True true :)
